domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt

```diff
@@ -0,0 +1,291 @@
+private const val DELETE_EXPLORATION_CHECKPOINT_DATA_PROVIDER_ID =
+  "delete_exploration_checkpoint_provider_id"
+
+/**
+ * Controller for saving, retrieving, updating, and deleting exploration checkpoints.
+ */
+@Singleton
+class ExplorationCheckpointController @Inject constructor(
+  private val cacheStoreFactory: PersistentCacheStore.Factory,
+  private val dataProviders: DataProviders,
+  private val oppiaLogger: OppiaLogger,
+  @ExplorationStorageDatabaseSize private val explorationCheckpointDatabaseSizeLimit: Int
+) {
+
+  /** Different stages in which the exploration checkpoint database can exist. */
+  enum class ExplorationCheckpointDatabaseState {
+
+    /** checkpoint database has not exceeded the allocated size limit. */
+    CHECKPOINT_DATABASE_SIZE_LIMIT_NOT_EXCEEDED,
+
+    /** checkpoint database has exceeded the allocated size limit. */
+    CHECKPOINT_DATABASE_SIZE_LIMIT_EXCEEDED
+  }
+
+  /** Indicates that no checkpoint was found for the specified explorationId and profileId. */
+  class ExplorationCheckpointNotFoundException(message: String) : Exception(message)
+
+  /**
+   * These Statuses correspond to the exception and the checkpoint database states above
+   * such that if the deferred result contains
+   *
+   * CHECKPOINT_SAVED_DATABASE_SIZE_LIMIT_EXCEEDED, the
+   * [ExplorationCheckpointDatabaseState.CHECKPOINT_DATABASE_SIZE_LIMIT_EXCEEDED] will be
+   * passed to a successful AsyncResult.
+   *
+   * CHECKPOINT_SAVED_DATABASE_SIZE_LIMIT_NOT_EXCEEDED,
+   * [ExplorationCheckpointDatabaseState.CHECKPOINT_DATABASE_SIZE_LIMIT_NOT_EXCEEDED] will be
+   * passed to a successful AsyncResult.
+   *
+   * CHECKPOINT_NOT_FOUND, the [ExplorationCheckpointNotFoundException] will be passed to a failed
+   * AsyncResult.
+   *
+   * SUCCESS corresponds to successful AsyncResult with value as null.
+   */
+  enum class ExplorationCheckpointActionStatus {
+    CHECKPOINT_SAVED_DATABASE_SIZE_LIMIT_NOT_EXCEEDED,
+    CHECKPOINT_SAVED_DATABASE_SIZE_LIMIT_EXCEEDED,
+    CHECKPOINT_NOT_FOUND,
+    SUCCESS
+  }
+
+  private val cacheStoreMap =
+    mutableMapOf<ProfileId, PersistentCacheStore<ExplorationCheckpointDatabase>>()
+
+  /**
+   * Records an exploration checkpoint for the specified profile.
+   *
+   * @return a [DataProvider] that indicates the success/failure of the save operation.
+   *
+   * Success result is returned if the checkpoint is saved successfully. The success result
+   * returned has the value
+   * [ExplorationCheckpointDatabaseState.CHECKPOINT_DATABASE_SIZE_LIMIT_EXCEEDED]
+   * if the database has exceeded the size limit of [explorationCheckpointDatabaseSizeLimit],
+   * otherwise the success result is returned with the value
+   * [ExplorationCheckpointDatabaseState.CHECKPOINT_DATABASE_SIZE_LIMIT_NOT_EXCEEDED].
+   */
+  fun recordExplorationCheckpoint(
+    profileId: ProfileId,
+    explorationId: String,
+    explorationCheckpoint: ExplorationCheckpoint
+  ): DataProvider<Any?> {
+    val deferred =
```

Comment:
Could you use more specific variable names here and elsewhere

Reply:
This naming makes the variable names consistent with other controllers.

Reply:
Okay makes sense